### ISSUE TITLE
Timeout issue

### ISSUE THREAD

**Issue by eltesco1** [Jamal135/IndeedScraper#2]
Hi Jamal

I am getting errors as below. I Have a solid and fast internet connection but it seems to fail after scraping a few items (number changes each time). 

I have tried to separate internet connections in two locations.

Since the items changes each time i think it may just need an increase to the timeout timer?

See below:

C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main>launcher.py
[-] Preparing to launch scraper...
[i] Generated output name: result_0
[?] Please select a scraper configuration: test.json

test.json

[-] Calling scraper launcher...
[i] Running Selenium driver without header...
[i] Disabled Selenium driver logging...
[i] Loaded test.json contents:
DXC Technology: https://www.seek.com.au/companies/dxc-technology-797000/reviews
[-] Scraping DXC Technology
**16it** [01:15, 4.74s/it]  **##This changes each time it fails - somtimes 1it, sometime 5it etc.**
[!] ALERT: Error occurred, ending Selenium driver session...
[!] ERROR: Unexpected error: could be internet...
scraper:Seek TimeoutException
Message:

[!] TRACE:
File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 101, in scrape_launch
scrape_website(driver, scraper, config)
File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 88, in scrape_website
data_blocks = scrape_data(driver, scraper, config)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 72, in scrape_data
SE.handle_bad_data(scraper.navigators.wait_for_page, config.data_strict, driver)
File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\custom_exceptions.py", line 35, in handle_bad_data
func(*args, **kwargs)
File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\scrapers\Seek.py", line 133, in wait_for_page
wait.until(page_has_changed)
File "C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\selenium\webdriver\support\wait.py", line 95, in until
raise TimeoutException(message, screen, stacktrace)

[*] DUMP:
{'data_strict': True,
'orgs': [{'name': 'DXC Technology',
'url': 'https://www.seek.com.au/companies/dxc-technology-797000/reviews'}],
'output_name': 'result_0',
'scraper_name': 'Seek',
'selenium_header': False,
'selenium_logging': False}

**Comment by eltesco1** [comment]
ALso, when it got a little further i got this error which seems to be a little different?

[-] Calling scraper launcher...
[i] Running Selenium driver without header...
[i] Disabled Selenium driver logging...
[i] Loaded test.json contents:
DXC Technology: https://www.seek.com.au/companies/dxc-technology-797000/reviews
[-] Scraping DXC Technology
24it [01:22,  3.43s/it]
[!] ALERT: Error occurred, ending Selenium driver session...
[!] ERROR: Unexpected error: could be internet...
scraper:Seek StaleElementReferenceException
Message: stale element reference: stale element not found
  (Session info: headless chrome=117.0.5938.63); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#stale-element-reference-exception
Stacktrace:
        GetHandleVerifier [0x002DCE43+45171]
        (No symbol) [0x00269091]
        (No symbol) [0x0015BE1D]
        (No symbol) [0x0015F9A7]
        (No symbol) [0x00160D6A]
        (No symbol) [0x00160E10]
        (No symbol) [0x0018B91D]
        (No symbol) [0x001AB04C]
        (No symbol) [0x00187FA6]
        (No symbol) [0x001AB2C4]
        (No symbol) [0x001BDCFA]
        (No symbol) [0x001AAE46]
        (No symbol) [0x0018675E]
        (No symbol) [0x001878FD]
        GetHandleVerifier [0x00595449+2897529]
        GetHandleVerifier [0x005DDBFB+3194411]
        GetHandleVerifier [0x005D79D1+3169281]
        GetHandleVerifier [0x003660C0+606960]
        (No symbol) [0x00272E0C]
        (No symbol) [0x0026EF98]
        (No symbol) [0x0026F0BF]
        (No symbol) [0x002617B7]
        BaseThreadInitThunk [0x772BFCC9+25]
        RtlGetAppContainerNamedObjectPath [0x77B27B1E+286]
        RtlGetAppContainerNamedObjectPath [0x77B27AEE+238]
        (No symbol) [0x00000000]

[!] TRACE:
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 101, in scrape_launch
    scrape_website(driver, scraper, config)
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 88, in scrape_website
    data_blocks = scrape_data(driver, scraper, config)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\scraper_controller.py", line 72, in scrape_data
    SE.handle_bad_data(scraper.navigators.wait_for_page, config.data_strict, driver)
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\utilities\custom_exceptions.py", line 35, in handle_bad_data
    func(*args, **kwargs)
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\scrapers\Seek.py", line 125, in wait_for_page
    old_texts = [elem.text for elem in driver.find_elements(By.TAG_NAME, 'h3')]
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Downloads\pyscrapify-main (1)\pyscrapify-main\scrapers\Seek.py", line 125, in <listcomp>
    old_texts = [elem.text for elem in driver.find_elements(By.TAG_NAME, 'h3')]
                 ^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 89, in text
    return self._execute(Command.GET_ELEMENT_TEXT)["value"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 394, in _execute
    return self._parent.execute(command, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 344, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)

[*] DUMP:
{'data_strict': True,
 'orgs': [{'name': 'DXC Technology',
           'url': 'https://www.seek.com.au/companies/dxc-technology-797000/reviews'}],
 'output_name': 'result_0',
 'scraper_name': 'Seek',
 'selenium_header': False,
 'selenium_logging': False}

**Comment by McWashr** [comment]
Your first error dump is a TimeoutException for sure, second one being a StaleElementReferenceException likely also indicates a page that was expected to load didn't... I was able to rule out it being to do with the scraper itself by testing DXC Technology. settings.py has a webdriver timeout constant, set to 20s by default. I would recommend testing with a higher value on that constant. I didn't make this a setting yet but in scraper_controller.py, the final launch scraper function, set selenium_header to True, will help you see what is happening when it breaks.

**Comment by eltesco1** [comment]
Hi Jamal,

Sorry to be difficult but unable to get any to complete. Ive tried two computers, two different internet connections, increasing timeout etc. DXC always fails on item 24 now, target fails on around 34. Could it be about chrome settings? of OS on PC (eg. Windows vs. Mac)?

**Comment by eltesco1** [comment]
OK, I think i have sorted it!

Windows Language seeting must match the seek country. I had to change my language settings in Windows to be Australian instead of US which was the default.

Hope this helps others.

**Comment by McWashr** [comment]
Cheers Eltesco1, really appreciate this update. I will treat this as a bug as I suspect I can enforce browser language and make it a settings.py constant. Will close pending a resolution I can find.

**Comment by McWashr** [comment]
Pending the next release I have updated the BaseScraper BaseParsers class to expect a browser_lang code which is passed into selenium_handler when a Chrome browser session is created. I have also updated the Seek scraper to specify browser_language as Australian English. Here is a link to the best list of browser language codes I could find for future reference: https://cloud.google.com/speech-to-text/docs/languages 

The issue was resolved this way given the original approach of making browser language a settings.py constant would have forced the assumption everyone that uses this framework wants to only scrape with Australian English. The challenge of diagnosing this issue reflects a lack of effective error detection and handling - this is planned to also be resolved in next release. Issue closed, cheers.
